src/util/mp_arith.cpp

```diff
@@ -258,8 +258,8 @@ mp_integer arith_left_shift(
 {
   PRECONDITION(a.is_long() && b.is_ulong());
   ullong_t shift=b.to_ulong();
-  if(shift>true_size && a!=mp_integer(0))
-    throw "shift value out of range";
+  DATA_INVARIANT(!(shift>true_size && a!=mp_integer(0)),
```

Comment:
I'd expand the not here:

`shift <= true_size || a == 0`

I'd also put it above as a precondition (definitely not data invariant, mp_integers can have all sorts of values that violate these conditions):

`PRECONDITION(b <= true_size || a == 0)`

The explicit conversion to `mp_integer` isn't needed...

Also, not part of this PR but these parameters could use better names than `a` and `b`...